doloop.py

```diff
@@ -686,25 +689,40 @@ def bump(dbconn, table, id_or_ids, lock_for=0, auto_add=True, test=False):
     """
     _check_table_is_a_string(table)
 
-    if not isinstance(lock_for, (int, long, float)):
-        raise TypeError('lock_for must be a number, not %r' %
-                        (lock_for,))
+    lock_for_values = _to_list(lock_for)
+    if not (1 <= len(lock_for_values) <= 2 and all(
+        isinstance(lock_for_value, (int, long, float))
+        for lock_for_value in lock_for_values
+    )):
+        raise TypeError('lock_for must be a number or 2-tuple of numbers, not'
+                        '%r' % (lock_for,))
+
+    if len(lock_for_values) == 1:
+        lock_for_sql = '?'
+        params = [lock_for]
+    else:
+        min_lock_for, max_lock_for = lock_for_values
+        if min_lock_for > max_lock_for:
+            raise ValueError('min_lock_for must not exceed max_lock_for')
+        lock_for_sql = 'FLOOR(? + RAND() * (? - ?))'
+        params = [min_lock_for, max_lock_for, min_lock_for]
 
     ids = _to_list(id_or_ids)
     if not ids:
         return 0
+    params += ids
 
-    sql = ('UPDATE `%s` SET `lock_until` = UNIX_TIMESTAMP() + ?'
+    sql = ('UPDATE `%s` SET `lock_until` = UNIX_TIMESTAMP() + @lock_for := %s'
```

Comment:
Looking at the MySQL documentation, it might not be the best thing that I'm assigning and referencing the `@lock_for ` in the same query.
>As a general rule, other than in SET statements, you should never assign a value to a user variable and read the value within the same statement

I'm not sure of a better way to ensure that the `lock_until` value for each row is strictly decreasing with a single update query. I could provide a randomly generated seed to RAND() to guarantee that the value generated by RAND() for each row is the same with multiple calls. Could also have the whereclause be conservative with `lock_until > UNIX_TIMESTAMP() + <max_lock_for>`.